### PR TITLE
arch & board: Add Espressif ESP32 board code.

### PR DESCRIPTION
**Pull request by xiarihf**
Add Espressif ESP32 board code.

Signed-off-by: xiarihf qiang3.zhang@samsung.com

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [1e2285d98465471da27eb0f1b8f45fa3fc0daca8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1e2285d98465471da27eb0f1b8f45fa3fc0daca8] -

**Comment by seinfra** [comment]
Target : [9d2109360c52a496c7d6093b95f320b866bdd3c6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [9d2109360c52a496c7d6093b95f320b866bdd3c6] -

**Comment by xiarihf** [comment]
@samirsahu1947 @sunghan-chang @manoharahk @yashwanth686007 
Dear Sir.
     ESP32 chip and ESP32 Devkit board code have been submitted, but timeout occurs in the crc-c check.  I have check the log, but has no idea. 
     Can you give us any suggestion?  Thank you very much.  

Best wishes,
Qiang.

File: build/configs/esp32_DevKit/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+## Contents
+
+> [Information](#information)  
+> [Environment Set-up](#environment-set-up)  
+ 
+## Information
+
+will be updated
+
+## Environment Set-up
```
Comment (xixidodo) [inline build/configs/esp32_DevKit/README.md]: change the title to  'Manual Environment Set-up', and leave a comment like 'docker method will be updated later '

File: build/configs/esp32_DevKit/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+This command retrieves the specific pre-configured file named defconfig and Make.defs according to \<board\>/\<configuration_set\>.  
+You can see collection of all configuration files at *$TIZENRT_BASEDIR/build/configs*.  
+To check all pre-defined configurations, type as follows:
+
+```bash
+./configure.sh --help
+```
+
+After configuring above, configuration can be modified through *make menuconfig* from *$TIZENRT_BASEDIR/os*.
+Howerver, menuconfig isn't ready for Xtensa architecture esp32 board. Please do not use it for esp32 board.
```
Comment (xixidodo) [inline build/configs/esp32_DevKit/README.md]: I think you have added menuconfig feature already.

File: os/arch/xtensa/include/esp32/rtc_io_struct.h [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+				uint32_t sense3_fun_sel :2;	/*the functional selection signal of the pad */
+				uint32_t sense2_fun_ie :1;	/*the input enable of the pad */
+				uint32_t sense2_slp_ie :1;	/*the input enable of the pad in sleep status */
+				uint32_t sense2_slp_sel :1;	/*the sleep status selection signal of the pad */
+				uint32_t sense2_fun_sel :2;	/*the functional selection signal of the pad */
+				uint32_t sense1_fun_ie :1;	/*the input enable of the pad */
+				uint32_t sense1_slp_ie :1;	/*the input enable of the pad in sleep status */
+				uint32_t sense1_slp_sel :1;	/*the sleep status selection signal of the pad */
+				uint32_t sense1_fun_sel :2;	/*the functional selection signal of the pad */
+				uint32_t sense4_mux_sel :1;	/*�1� select the digital function  �0�slection the rtc function */
```
Comment (xixidodo) [inline os/arch/xtensa/include/esp32/rtc_io_struct.h]: check encoding format and content in the comments

File: os/arch/xtensa/Kconfig [hunk 4]
```diff
@@ -5,13 +5,41 @@
...
 config ARCH_FAMILY_LX6
 	bool
 	default n
 	---help---
 		Cadence® Tensilica® Xtensa® LX6 data plane processing unit (DPU).
 		The LX6 is a configurable and extensible processor core.
 
+config ARCH_CHIP
+	string
+	default "esp32"		if ARCH_CHIP_ESP32
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply os/arch/xtensa/Kconfig]: no problem, :0

File: build/configs/esp32_DevKit/README.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# esp32_DevKit
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/configs/esp32_DevKit/README.md]: fixed

File: build/configs/esp32_DevKit/README.md (hunk 1 above)
Comment (xiarihf) [reply build/configs/esp32_DevKit/README.md]: fixed, thanks.

File: build/configs/esp32_DevKit/README.md (hunk 2 above)
Comment (xiarihf) [reply build/configs/esp32_DevKit/README.md]: fixed, thanks.

File: os/arch/xtensa/include/esp32/rtc_io_struct.h (hunk 3 above)
Comment (xiarihf) [reply os/arch/xtensa/include/esp32/rtc_io_struct.h]: fixed, thanks.

**Comment by seinfra** [comment]
Target : [8ea64d12f564d21e60d199bafc8a73c0dd0a6e93] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8ea64d12f564d21e60d199bafc8a73c0dd0a6e93] -

**Comment by seinfra** [comment]
Target : [1805fe2c3f5700c9d1a51cdebd9f150eedfcca8d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1805fe2c3f5700c9d1a51cdebd9f150eedfcca8d] -

**Comment by seinfra** [comment]
Target : [54a1f292c5381313871f432a989a4cc51afbec2a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [54a1f292c5381313871f432a989a4cc51afbec2a] -

File: os/board/esp32_DevKitC/src/esp32_boot.c [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+	board_i2c_initialize();
+	board_ledc_setup();
+
+#if defined(CONFIG_ADC)
+	board_adc_initialize();
+#endif
+
+/*Init SPI RAM*/
+#ifdef CONFIG_SPIRAM_SUPPORT
+    esp_spiram_init_cache();
```
Comment (sunghan-chang) [inline os/board/esp32_DevKitC/src/esp32_boot.c]: Please check violations of coding rules.
Indentation (start of line) should be a tab, not spaces.

File: os/board/esp32_DevKitC/src/esp32_bringup.c [hunk 7]
```diff
@@ -0,0 +1,107 @@
+/******************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline os/board/esp32_DevKitC/src/esp32_bringup.c]: Please set a year 2019, not 2018.

Comment (sunghan-chang) [reply os/board/esp32_DevKitC/src/esp32_bringup.c]: For all of files

File: os/board/esp32_DevKitC/src/esp32_led.c [hunk 8]
```diff
@@ -0,0 +1,39 @@
+#include <sys/types.h>
```
Comment (sunghan-chang) [inline os/board/esp32_DevKitC/src/esp32_led.c]: Here is no license. Please add.

File: os/board/esp32_DevKitC/src/esp32_led.c [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+//#include <tinyara/ledc.h>
+#include "esp32_ledc.h"
+
+FAR struct ledc_lowerhalf_s *esp32_ledcinitialize(int i);
+
+int board_ledc_setup(void)
+{
+#ifdef CONFIG_LEDC
+    FAR struct ledc_lowerhalf_s *ledc;
```
Comment (sunghan-chang) [inline os/board/esp32_DevKitC/src/esp32_led.c]: Please check indentation. a tab

File: os/board/esp_wrover_kit/Kconfig [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+if ARCH_BOARD_ESP_WROVER_KIT
+
+choice
+	prompt "On-board Crystal Frequency"
+	default ESP32CORE_XTAL_40MZ
+
+config ESP32CORE_XTAL_40MZ
+	bool "40MHz"
+
+config ESP32CORE_XTAL_26MHz
```
Comment (sunghan-chang) [inline os/board/esp_wrover_kit/Kconfig]: 1. are there two kinds of ESP WROVER KIT? One has 40MHz and another has 26MHz?
    Or does it determine working freq.? Is there no way to change it on run-time? Is it static?
2. 40MZ -> 40MHz

File: os/board/esp_wrover_kit/src/esp32_boot.c [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+    esp_spiram_init_cache();
+    esp_spiram_init();
+#endif
+
+/*no-os heap configure*/
+#ifdef CONFIG_SPIRAM_USE_CAPS_ALLOC
+   // heap_caps_init();
+    esp_spiram_add_to_heapalloc();
+    //Assgin DMA memory for drivers*/
+    uint8_t *dma_heap = malloc(CONFIG_SPIRAM_MALLOC_RESERVE_INTERNAL);
```
Comment (jeongchanKim) [inline os/board/esp_wrover_kit/src/esp32_boot.c]: You'd better do type casting after malloc.

File: os/board/esp_wrover_kit/src/esp32_led.c [hunk 12]
```diff
@@ -0,0 +1,39 @@
+#include <sys/types.h>
```
Comment (jeongchanKim) [inline os/board/esp_wrover_kit/src/esp32_led.c]: Please add license

File: os/board/esp32_DevKitC/src/esp32_boot.c (hunk 6 above)
Comment (xiarihf) [reply os/board/esp32_DevKitC/src/esp32_boot.c]: Fixed, thanks.

File: os/board/esp32_DevKitC/src/esp32_bringup.c (hunk 7 above)
Comment (xiarihf) [reply os/board/esp32_DevKitC/src/esp32_bringup.c]: OK, all the new files in this commit are fixed, thanks.

File: os/board/esp32_DevKitC/src/esp32_led.c (hunk 8 above)
Comment (xiarihf) [reply os/board/esp32_DevKitC/src/esp32_led.c]: Fixed, thanks.

File: os/board/esp32_DevKitC/src/esp32_led.c (hunk 9 above)
Comment (xiarihf) [reply os/board/esp32_DevKitC/src/esp32_led.c]: Fixed ,thanks.

File: os/board/esp_wrover_kit/src/esp32_led.c (hunk 12 above)
Comment (xiarihf) [reply os/board/esp_wrover_kit/src/esp32_led.c]: fixed, thanks.

File: os/board/esp_wrover_kit/src/esp32_boot.c (hunk 11 above)
Comment (xiarihf) [reply os/board/esp_wrover_kit/src/esp32_boot.c]: yes, fixed, thanks.

**Comment by seinfra** [comment]
Target : [427f7cb4c3a256bce8993024d869a96704746214] -

**Comment by seinfra** [comment]
Target : [427f7cb4c3a256bce8993024d869a96704746214] -

**Comment by seinfra** [comment]
Target : [760b422a6b3247a03e154740a40b913116bb4b65] - Code Rule Check (C++) OK.

File: os/board/esp_wrover_kit/Kconfig (hunk 10 above)
Comment (xiarihf) [reply os/board/esp_wrover_kit/Kconfig]: First,  i will modify the configuration name. 
Secordly, for XTAL and CPU frequency, so far it's set in bootloader and not be implemented in ESP32 chip code.  Now this branch is only for bringing up esp32 board. We will add cpu frequency settings later.
Thanks.

**Comment by sunghan-chang** [comment]
@xiarihf CRC Server has a problem itself, not in your codes. CI Server maintainer is looking for this. Sorry for inconvience.

**Comment by sunghan-chang** [comment]
@xiarihf You should resolve a conflict.

File: os/board/esp32_DevKitC/src/esp32_bringup.c (hunk 7 above)
Comment (sunghan-chang) [reply os/board/esp32_DevKitC/src/esp32_bringup.c]: Not fixed yet. Please check it again.

**Comment by sunghan-chang** [comment]
@xiarihf CI maintainer said, it looks a network problem. Could you just update this PR?

File: os/board/esp32_DevKitC/Kconfig [hunk 13]
```diff
@@ -0,0 +1,204 @@
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+if ARCH_BOARD_ESP32_DEVKITC
+
+choice
+	prompt "On-board Crystal Frequency"
+	default ESP32CORE_XTAL_40MHZ
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply os/board/esp32_DevKitC/Kconfig]: Oh, yes, we can select the crystal. This function will be added later.

**Comment by xiarihf** [comment]
@sunghan-chang 
I will fixed the conflict. and other issues. 
Thanks.

**Comment by seinfra** [comment]
Target : [4883500f462eb6fb6d777b0c8274c92ff3c78d81] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a171d7b3e6a78083ee2c2743a1e878b8f2ba31de] - Code Rule Check (C++) OK.

**Comment by manoharahk** [comment]
Hello @xiarihf,

I could see following changes as part of ESP32 board port.
1.  Changes to arch/xtensa
2.  Add new board "esp32_DevKitC"
3.  Add new board "esp_wrover_kit"
4.  Create build scripts & "hello_with_tash" configuration  for esp32_DevKitC board
5.  Create build scripts & "hello_with_tash" configuration for esp32_DevKitC board

Is it possible to create a separate commit for all these changes?

**Comment by xiarihf** [comment]
> Hello @xiarihf,
> 
> I could see following changes as part of ESP32 board port.
> 
> 1. Changes to arch/xtensa
> 2. Add new board "esp32_DevKitC"
> 3. Add new board "esp_wrover_kit"
> 4. Create build scripts & "hello_with_tash" configuration  for esp32_DevKitC board
> 5. Create build scripts & "hello_with_tash" configuration for esp_wrover_kit board
> 
> Is it possible to create a separate commit for all these changes?

It's a good suggestion, thanks.

File: build/tools/esp32/Makefile [hunk 14]
```diff
@@ -0,0 +1,22 @@
+###########################################################################
+# 
+# Copyright 2018 Samsung Electronics All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/Makefile]: fixed, thanks.

File: build/tools/esp32/bootloader/Makefile [hunk 15]
```diff
@@ -0,0 +1,22 @@
+###########################################################################
+# 
+# Copyright 2018 Samsung Electronics All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/bootloader/Makefile]: fixed,thanks

File: build/tools/esp32/esptool_py/Makefile [hunk 16]
```diff
@@ -0,0 +1,153 @@
+###########################################################################
+# 
+# Copyright 2018 Samsung Electronics All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/esptool_py/Makefile]: fixed, thanks

File: build/tools/esp32/bootloader/Makefile [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+###########################################################################
+
+BOOTLOADER_BIN=$(CURRENTDIR)/bootloader/bootloader.bin
+BOOTLOADER_OFFSET := 0x1000
+
+ESPTOOL_BOOTLOADER_FLASH_ARGS := $(BOOTLOADER_OFFSET) $(BOOTLOADER_BIN)
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/bootloader/Makefile]: fixed, thanks.

File: build/tools/esp32/esptool_py/esptool.py [hunk 18]
```diff
@@ -0,0 +1,2867 @@
+###########################################################################
+# 
+# Copyright 2018 Samsung Electronics All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/esptool_py/esptool.py]: yes, all the file have been checked now. thanks.

File: build/tools/esp32/esptool_py/esptool.py [hunk 19]
```diff
@@ -0,0 +1,2867 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+###########################################################################
+
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/esptool_py/esptool.py]: yes, it's OK.

File: build/tools/esp32/partiton_table/Kconfig [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+    bool "Single factory app, no OTA"
+config PARTITION_TABLE_TWO_OTA
+    bool "Factory app, two OTA definitions"
+config PARTITION_TABLE_CUSTOM
+    bool "Custom partition table CSV"
+endchoice
+
+config PARTITION_TABLE_CUSTOM_FILENAME
+    string "Custom partition CSV file" if PARTITION_TABLE_CUSTOM
+    default partitions.csv
```
[the remark this answers is not in the document]
Comment (xiarihf) [reply build/tools/esp32/partiton_table/Kconfig]: fixed, all the string type config items.

**Comment by seinfra** [comment]
Target : [935f15b5fc5cedccfa45108d8feff48fc16dabd2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [935f15b5fc5cedccfa45108d8feff48fc16dabd2] -

File: build/tools/esp32/partiton_table/Kconfig [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+        Name of the custom partition CSV filename. This path is evaluated
+        relative to the project root directory.
+
+config PARTITION_TABLE_FILENAME
+       string
+       default "partitions_singleapp.csv" if PARTITION_TABLE_SINGLE_APP && !ESP32_ENABLE_COREDUMP_TO_FLASH
+       default "partitions_singleapp_coredump.csv" if PARTITION_TABLE_SINGLE_APP && ESP32_ENABLE_COREDUMP_TO_FLASH
+       default "partitions_two_ota.csv" if PARTITION_TABLE_TWO_OTA && !ESP32_ENABLE_COREDUMP_TO_FLASH
+       default "partitions_two_ota_coredump.csv" if PARTITION_TABLE_TWO_OTA && ESP32_ENABLE_COREDUMP_TO_FLASH
+       default "PARTITION_TABLE_CUSTOM_FILENAM"E if PARTITION_TABLE_CUSTOM
```
Comment (sunghan-chang) [inline build/tools/esp32/partiton_table/Kconfig]: M"E -> ME"

Comment (xiarihf) [reply build/tools/esp32/partiton_table/Kconfig]: oh, sorry, it's a mistake.

**Comment by seinfra** [comment]
Target : [c128496d0a7ac3005126b322583298bf36b78df3] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c128496d0a7ac3005126b322583298bf36b78df3] -

**Comment by sunghan-chang** [comment]
@xiarihf Could you spilt commits as @manoharahk suggested and as you agreed?

>> Hello @xiarihf,
>>
>> I could see following changes as part of ESP32 board port.
>>
>> Changes to arch/xtensa
>> Add new board "esp32_DevKitC"
>> Add new board "esp_wrover_kit"
>> Create build scripts & "hello_with_tash" configuration for esp32_DevKitC board
>> Create build scripts & "hello_with_tash" configuration for esp_wrover_kit board
>> Is it possible to create a separate commit for all these changes?
>
> It's a good suggestion, thanks.

**Comment by sunghan-chang** [comment]
@xiarihf To split them, just refer belows.
```
git reset --soft HEAD^
git reset HEAD .
git add os/arch/
git commit (1st commit for arch)

git add os/board/<boardA folder>
git commit (2nd commit for board A)

git add os/board/<boardB folder>
git commit (3rd commit for board B)

... for configA folder ...
... for configB folder ...

git push -f
```

**Comment by xiarihf** [comment]
> @xiarihf Could you spilt commits as @manoharahk suggested and as you agreed?
> 
> > > Hello @xiarihf,
> > > I could see following changes as part of ESP32 board port.
> > > Changes to arch/xtensa
> > > Add new board "esp32_DevKitC"
> > > Add new board "esp_wrover_kit"
> > > Create build scripts & "hello_with_tash" configuration for esp32_DevKitC board
> > > Create build scripts & "hello_with_tash" configuration for esp_wrover_kit board
> > > Is it possible to create a separate commit for all these changes?
> > 
> > 
> > It's a good suggestion, thanks.

@sunghan-chang , Yes, I'm trying the steps.  I will seperate this PR and make new commits, when there is no mistake in my git operations.

**Comment by seinfra** [comment]
Target : [28048fefd6c81201a62ddbecfed7214f6cc2e998] - Code Rule Check (C++) OK.

**Pull request merged**